File: Python/651-700/FindDuplicateSubtrees.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def findDuplicateSubtrees(self, root):
        """
        :type root: TreeNode
        :rtype: List[TreeNode]
        """
        from collections import defaultdict
        m = defaultdict(int)
        res = []
        def to_s(root):
            if root == None:
                return 'null'
            s = str(root.val)
            if root.left != None or root.right != None:
                s += ',' + to_s(root.left) + ',' + to_s(root.right)
            if m[s] == 1:
                res.append(root)
            m[s] += 1
            return s
        to_s(root)
        return res

File: Java/651-700/FindDuplicateSubtrees.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class FindDuplicateSubtrees {
    public List<TreeNode> findDuplicateSubtrees(TreeNode root) {
        List<TreeNode> res = new ArrayList<>();
        run(root, new HashMap<>(), res);
        return res;
    }
    String run(TreeNode root, Map<String, Integer> map, List<TreeNode> res) {
        if (root == null) return "null";
        String s = root.val + "," + run(root.left, map, res) + "," + run(root.right, map, res);
        if (map.containsKey(s)) {
            if (map.get(s) == 1) res.add(root);
            map.put(s, map.get(s) + 1);
        } else map.put(s, 1);
        return s;
    }
}
